screws_tilt_adjust: Add maximum deviation parameter and check

driest opened the pull request:
Add the `max_deviation` config parameter that is useful when a saved bed mesh is used,
to ensure that the bed level has not drifted too far from where it was when
the mesh was created. `SCREWS_TILT_CALCULATE` can be added to the custom start
gcode of the slicer before the mesh is loaded. It will pause the print with a
message asking the user to adjust the bed screws, after which it can be continued
with the `RESUME` command.

Signed-off-by: Johannes Stuettgen <johannes.stuettgen@gmail.com>

KevinOConnor commented:
Thanks.  However, I'm not sure running `PAUSE` is a good idea from a utility like this.  The user may not have PAUSE defined.

Separately, FWIW, I think it may make sense to change the screws_tilt_adjust module to follow the same process that z_tilt and quad_gantry_level use.  Including reusing the same helper code.

@rmcbc - FYI.

-Kevin

driest commented:
Is there a better way to handle interactive adjustments by the user? I've looked at RetryHelper in z_tilt, but this seems non interactive. Alternatively one could also abort the print with an error, but that has the disadvantage that it will disable all heaters and thus increase the overall time required for this process due to the bed and nozzle cooling down.

KevinOConnor commented:
I don't use the screws_tilt_adjust module, so I'm not sure what actions you'd like to achieve.  Is the goal to run the probe measurements, allow the user to make adjustments, and then run the probe measurements again to see if more adjustments are needed?

-Kevin

driest commented:
exactly. i don't have a self leveling bed, but would like to load a stored bed mesh without having to measure all probe points every print.

With the changes in this pull request I can simply add the screw tilt adjust command to my start gcode, and instead of probing the entire mesh the printer will just probe the 4 bed adjustment screws. If the bed level exceeds a configured threshold it will pause the print and tell me exactly how far to turn each bed screw to get back into alignment. After adjusting the screws I can then manually run the screw tilt adjust module again to verify my adjustments were correct, and then resume the print.

Basically this creates an invariant that allows the user to treat the bed as physically level, and briefly enters an interactive adjustment mode if it exceeds the configured threshold.

The pause function is critical for this, as it greatly simplifies the adjustment process. It would be possible to achieve the same goal by simply triggering a shutdown if the configured limit is exceeded, but that would cause the heaters to shut down and thus prolong the adjustment process and cause inconsistent measurements due to temperature variations during the adjustment process.





\-------- Original Message --------
On Feb 19, 2021, 03:44, KevinOConnor < notifications@github.com> wrote:

>
>
>
> I don't use the screws\_tilt\_adjust module, so I'm not sure what actions you'd like to achieve. Is the goal to run the probe measurements, allow the user to make adjustments, and then run the probe measurements again to see if more adjustments are needed?
>
> \-Kevin
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [view it on GitHub][], or [unsubscribe][].![AAD6WD6454ELAKU5VDLUQULS7XGCVA5CNFSM4XJKOAIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOF2MPDVQ.gif][]


[view it on GitHub]: https://github.com/KevinOConnor/klipper/pull/3904#issuecomment-781775318
[unsubscribe]: https://github.com/notifications/unsubscribe-auth/AAD6WD6LI2SGCKKT3UYBWG3S7XGCVANCNFSM4XJKOAIA
[AAD6WD6454ELAKU5VDLUQULS7XGCVA5CNFSM4XJKOAIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOF2MPDVQ.gif]: https://github.com/notifications/beacon/AAD6WD6454ELAKU5VDLUQULS7XGCVA5CNFSM4XJKOAIKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOF2MPDVQ.gif

KevinOConnor commented:
Okay.  Unless I'm missing something, though, it's possible to do that with a macro.  A macro could run PROBE at the four points, check if the z height is outside of some bound at each point, and invoke `{action_raise_error("bed not level")}` if so.  Raising an error is not a shutdown - it will just cancel the current command and/or current print.

-Kevin

driest commented:
You mean replacing the entire screw_tilt_adjust module with a macro? Because leaving it in and creating a macro just for interactive mode would mean:
a) replicating a lot of the modules' functionality in a macro
b) duplicating configuration in two places (screw positions, probe speeds, etc)

which seems less than ideal to me, but I don't have much experience with the macro language, so maybe I'm wrong.

KevinOConnor commented:
Earlier you wrote:

> I can simply add the screw tilt adjust command to my start gcode, and instead of probing the entire mesh the printer will just probe the 4 bed adjustment screws. If the bed level exceeds a configured threshold it will pause the print

It sounded like you were looking for a way to confirm the bed level prior to each print.  That confirmation of bed level is possible with a macro.  If there is an error, fixing it would still need screws_tilt_adjust.py .

Maybe I'm missing something though.

-Kevin

driest commented:
I certainly believe you when you say that this is possible with a macro, I just meant that this macro would basically replicate most of the code in screw_tilt_adjust, and require the duplication of configuration options such as screw positions. Does this seem cleaner to you than adding this simple check here in the screw_tilt_adjust module?

Regarding calling the PAUSE macro, I agree that this is not ideal if there are no guarantees that this will be present. How about I raise printer.command_error instead? that should have the same effect as {action_raise_error("bed not level")} from a macro if I understand correctly.

KevinOConnor commented:
Okay, I think I understand the goal of this change.  (Effectively, to implement a manual "retry" system like is done for z_tilt and quad_gantry_level.)  However, my high-level feedback is that I don't think the implementation here isn't a good candidate for the master Klipper branch.  (It issues a PAUSE command and differs from the way z_tilt and quad_gantry_level implement retries.)

-Kevin

driest commented:
Yes, exactly this is the goal. I've been using this patch on my Ender 3 v2 for 2 months now and really enjoy the workflow, because I basically never have to worry about leveling the bed. The printer will tell me when I try to print on a tilted bed, and give me a chance to fix it before resuming the print.

I had a look at z_tilt and quad_gantry_level, and the way retries are handled there is not applicable to this use case because here we need user interaction. RetryHelper in the two mentioned modules simply acts as a manager for automatic adjustments, but what is needed here is a moderator for user interaction.

I'm happy to factor the retry system out into a helper class to be more in line with the way z_tilt and quad_gantry_level do things, however I still need a way to interact with the user to achieve the following things:

1. Interrupt the printing process without shutting down the heating elements. This is necessary to avoid delays caused by waiting for bed and nozzle to come back up to temperature.
2. Report feedback to the user to guide him in adjusting the bed screws
3. Allow the user to issue gcode commands without affecting the following print. This is needed to allow the user to issue another SCREWS_TILT_ADJUST to verify his adjustments were correct, before resuming the print.

Please let me know if you have a better idea on how to achieve 1-3 without issuing PAUSE, and I'll port my code to whatever you suggest. Alternatively I'm also happy with a solution that stops the print with a message while not shutting down the heaters. As a last resort, I could also simply keep using PAUSE, but verify the PAUSE module is present before I use it.

KevinOConnor commented:
> I had a look at z_tilt and quad_gantry_level, and the way retries are handled there is not applicable to this use case because here we need user interaction.

Well, both z_tilt and quad_gantry_level are designed to work with user interaction.  Both use the ProbePointsHelper() class which works with either manual bed probing or automatic bed probing.

I think what you're saying is neither of those commands are intended to be called at the start of every print when in manual mode.  That is true.

> Allow the user to issue gcode commands without affecting the following print.

Alas, I don't know of a way to reliably do that with OctoPrint (or with a similar external program that feeds gcode commands via the pseudo-serial port).  The problem is that we have no way of preventing OctoPrint from sending commands before we've even finished the screws_tilt_adjust.  That is, we can't pause a command that's already been sent to Klipper.

I suspect this would work okay if it was in a print starting from virtual_sdcard, because in that case we know that Klipper will never try to process the next command without first fully processing the proceeding command.  In the general case, I don't know of a good way to reliably pause from within a print.  I fear processing some follow up commands and then pausing would lead to startling and confusing results.

> Please let me know if you have a better idea on how to achieve 1-3 without issuing PAUSE

FWIW, an alternate solution would be to introduce a new "verify_bed_level.py" extras module that checks for level and raises an error if the bed isn't level.  Reporting an error will stop the print, but it will not turn off the heaters.  (OctoPrint may be configured to disable the heaters on an error, but a user can customize that.)

-Kevin

Sineos commented:
I think this is an excellent feature and does match with my workflow perfectly. Instead of self-leveling, my bed is rather self-devastating so my typical workflow is:

1. Heat the bed to target temp
2. Run `SCREWS_TILT_CALCULATE`
    * If NOK level until within +/- 3 min
3. Start printing

Automating this would be awesome.

I noticed that `max_deviation` is given as float. Would it better match to have it in minutes to reflect the `SCREWS_TILT_CALCULATE` measurement?

driest commented:
I am using mainsail, and with the code in this pull request I can simply issue SCREWS_TILT_CALCULATE as often as needed after the module pauses the print, and then RESUME to continue printing. However, you make a great point that other web interfaces might not handle this as graciously, so erroring out is probably the best way to deal with this. I've changed the code to simply raise a gcode error when the bed level exceeds its configured limit. This integrates into my workflow just as well and doesn't cause the same problems as PAUSE.

The problem with refactoring this functionality into a separate module is that we'd have to replicate the entire configuration and logic of screws_tilt_adjust, as we need to know where the screws are to probe them, and what thread size they are to give the user any meaningful information on how to adjust. And I don't think duplicating configuration would be a great idea in this case.

Regarding units of measurement, I'd like to keep the max deviation in millimeters, to be able to quickly put that number into context with layer height as thats the most relevant metric when deciding how much of an error to tolerate (I'm keeping mine at 1/2 layer height with good results atm).

KevinOConnor commented:
Thanks.  In general it looks fine to me.  However, I'm not sure it makes sense to change the behaviour of the existing SCREWS_TILT_CALCULATE command.  Perhaps add a new `SCREWS_TILT_CHECK` command that takes a max_deviation parameter.  (Or, alternatively, perhaps add a max_deviation parameter to the existing SCREWS_TILT_CALCULATE command.

-Kevin

driest commented:
Apologies, Pycharms Github integration exploded, I'll fix it.

driest commented:
I have adapted the config parameter to a gcode argument, adjusted the documentation accordingly, and squashed everything into a single commit to undo the mess I've made and clean up the commit history. Please let me know if this matches your expectations or if there is anything else I can do to clean this up.

driest commented:
Also it seems that I accidently messed with the workflows, please just deny that request I can't seem to be able to cancel it from my side.

KevinOConnor commented:
Thanks.  I committed this.

FYI, it would also be good to add the MAX_DEVIATION parameter to docs/G-Codes.md .

-Kevin
